File: addons/guide-cpp/src/guide_input_mapping.cpp
#include "guide_input_mapping.h"
#include "triggers/guide_trigger_down.h"
#include "triggers/guide_trigger_hold.h"
#include <godot_cpp/core/math.hpp>
#include <algorithm>

using namespace godot;

void GUIDEInputMapping::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_override_action_settings"), &GUIDEInputMapping::get_override_action_settings);
    ClassDB::bind_method(D_METHOD("set_override_action_settings", "override"), &GUIDEInputMapping::set_override_action_settings);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "override_action_settings"), "set_override_action_settings", "get_override_action_settings");

    ClassDB::bind_method(D_METHOD("get_is_remappable"), &GUIDEInputMapping::get_is_remappable);
    ClassDB::bind_method(D_METHOD("set_is_remappable", "remappable"), &GUIDEInputMapping::set_is_remappable);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "is_remappable"), "set_is_remappable", "get_is_remappable");

    ClassDB::bind_method(D_METHOD("get_display_name"), &GUIDEInputMapping::get_display_name);
    ClassDB::bind_method(D_METHOD("set_display_name", "name"), &GUIDEInputMapping::set_display_name);
    ADD_PROPERTY(PropertyInfo(Variant::STRING, "display_name"), "set_display_name", "get_display_name");

    ClassDB::bind_method(D_METHOD("get_display_category"), &GUIDEInputMapping::get_display_category);
    ClassDB::bind_method(D_METHOD("set_display_category", "category"), &GUIDEInputMapping::set_display_category);
    ADD_PROPERTY(PropertyInfo(Variant::STRING, "display_category"), "set_display_category", "get_display_category");

    ClassDB::bind_method(D_METHOD("get_input"), &GUIDEInputMapping::get_input);
    ClassDB::bind_method(D_METHOD("set_input", "input"), &GUIDEInputMapping::set_input);
    ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "input", PROPERTY_HINT_RESOURCE_TYPE, "GUIDEInput"), "set_input", "get_input");

    ClassDB::bind_method(D_METHOD("get_modifiers"), &GUIDEInputMapping::get_modifiers);
    ClassDB::bind_method(D_METHOD("set_modifiers", "modifiers"), &GUIDEInputMapping::set_modifiers);
    ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "modifiers", PROPERTY_HINT_NONE, "24/17:GUIDEModifier", PROPERTY_USAGE_DEFAULT, "GUIDEModifier"), "set_modifiers", "get_modifiers");

    ClassDB::bind_method(D_METHOD("get_triggers"), &GUIDEInputMapping::get_triggers);
    ClassDB::bind_method(D_METHOD("set_triggers", "triggers"), &GUIDEInputMapping::set_triggers);
    ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "triggers", PROPERTY_HINT_NONE, "24/17:GUIDETrigger", PROPERTY_USAGE_DEFAULT, "GUIDETrigger"), "set_triggers", "get_triggers");
}

GUIDEInputMapping::GUIDEInputMapping() {
}

GUIDEInputMapping::~GUIDEInputMapping() {
}

void GUIDEInputMapping::_initialize(GUIDEAction::GUIDEActionValueType value_type) {
    _trigger_list.clear();
    _implicit_count = 0;
    _explicit_count = 0;
    _trigger_hold_threshold = -1.0;

    if (triggers.is_empty()) {
        Ref<GUIDETriggerDown> default_trigger;
        default_trigger.instantiate();
        default_trigger->set_actuation_threshold(0);
        _explicit_count = 1;
        _trigger_list.append(default_trigger);
        return;
    }

    Vector3 input_value = Vector3(0, 0, 0);
    if (input.is_valid()) {
        input_value = input->get_value();
    }

    for (int i = 0; i < modifiers.size(); i++) {
        Ref<GUIDEModifier> modifier = modifiers[i];
        if (modifier.is_valid()) {
            input_value = modifier->_modify_input(input_value, 0, value_type);
        }
    }

    for (int i = 0; i < triggers.size(); i++) {
        Ref<GUIDETrigger> trigger = triggers[i];
        if (!trigger.is_valid()) continue;

        switch (trigger->_get_trigger_type()) {
            case GUIDETrigger::EXPLICIT:
                _explicit_count++;
                break;
            case GUIDETrigger::IMPLICIT:
                _implicit_count++;
                break;
            case GUIDETrigger::BLOCKING:
                break;
        }
        _trigger_list.append(trigger);

        Ref<GUIDETriggerHold> hold_trigger = trigger;
        if (hold_trigger.is_valid()) {
            if (_trigger_hold_threshold == -1) {
                _trigger_hold_threshold = hold_trigger->get_hold_threshold();
            } else {
                _trigger_hold_threshold = Math::min(_trigger_hold_threshold, hold_trigger->get_hold_threshold());
            }
        }
        trigger->set_last_value(input_value);
    }
}

void GUIDEInputMapping::_update_state(double delta, GUIDEAction::GUIDEActionValueType value_type) {

    Vector3 input_value = Vector3(0, 0, 0);
    if (input.is_valid()) {
        input_value = input->get_value();
    }

    for (int i = 0; i < modifiers.size(); i++) {
        Ref<GUIDEModifier> modifier = modifiers[i];
        if (modifier.is_valid()) {
            input_value = modifier->_modify_input(input_value, delta, value_type);
            // UtilityFunctions::print("Modifier ", i, " processing input after: ", input_value);
        }
    }

    _value = input_value;

    int triggered_implicits = 0;
    int triggered_explicits = 0;
    int triggered_blocked = 0;

    int result = GUIDETrigger::NONE;

    for (int i = 0; i < _trigger_list.size(); i++) {
        Ref<GUIDETrigger> trigger = _trigger_list[i];
        if (!trigger.is_valid()) continue;

        GUIDETrigger::GUIDETriggerState trigger_result = trigger->_update_state(_value, delta, value_type);
        trigger->set_last_value(_value);

        GUIDETrigger::GUIDETriggerType trigger_type = trigger->_get_trigger_type();
        if (trigger_result == GUIDETrigger::TRIGGERED) {
            switch (trigger_type) {
                case GUIDETrigger::EXPLICIT:
                    triggered_explicits++;
                    break;
                case GUIDETrigger::IMPLICIT:
                    triggered_implicits++;
                    break;
                case GUIDETrigger::BLOCKING:
                    triggered_blocked++;
                    break;
            }
        }

        if (trigger_type == GUIDETrigger::EXPLICIT) {
            result = Math::max(result, (int)trigger_result);
        }
    }

    if (triggered_blocked > 0) {
        _state = GUIDETrigger::NONE;
        return;
    }

    if (triggered_implicits < _implicit_count) {
        _state = GUIDETrigger::NONE;
        return;
    }

    if (_explicit_count == 0 && _implicit_count > 0) {
        _state = GUIDETrigger::TRIGGERED;
        return;
    }

    _state = (GUIDETrigger::GUIDETriggerState)result;
}

File: addons/GuideCS/GUIDEInputMapping.cs
using System;
using System.Collections.Generic;
using Godot;

namespace Guide;

[Tool]
[GlobalClass, Icon("res://addons/GuideCS/GUIDEInternal.svg")]
public partial class GUIDEInputMapping : Resource {
	private bool _overrideActionSettings = false;
	[Export]
	public bool OverrideActionSettings {
		get => _overrideActionSettings;
		set {
			if (_overrideActionSettings == value) return;
			_overrideActionSettings = value;
			EmitChanged();
		}
	}

	private bool _isRemappable = false;
	[Export]
	public bool IsRemappable {
		get => _isRemappable;
		set {
			if (_isRemappable == value) return;
			_isRemappable = value;
			EmitChanged();
		}
	}

	private string _displayName = "";
	[Export]
	public string DisplayName {
		get => _displayName;
		set {
			if (_displayName == value) return;
			_displayName = value;
			EmitChanged();
		}
	}

	private string _displayCategory = "";
	[Export]
	public string DisplayCategory {
		get => _displayCategory;
		set {
			if (_displayCategory == value) return;
			_displayCategory = value;
			EmitChanged();
		}
	}

	[ExportGroup("Mappings")]
	private GUIDEInput _input;
	[Export]
	public GUIDEInput Input {
		get => _input;
		set {
			if (_input == value) return;
			_input = value;
			EmitChanged();
		}
	}

	private Godot.Collections.Array<GUIDEModifier> _modifiers = new();
	[Export]
	public Godot.Collections.Array<GUIDEModifier> Modifiers {
		get => _modifiers;
		set {
			if (_modifiers == value) return;
			_modifiers = value;
			EmitChanged();
		}
	}

	private Godot.Collections.Array<GUIDETrigger> _triggers = new();
	[Export]
	public Godot.Collections.Array<GUIDETrigger> Triggers {
		get => _triggers;
		set {
			if (_triggers == value) return;
			_triggers = value;
			EmitChanged();
		}
	}

	internal float _triggerHoldThreshold = -1.0f;
	internal GUIDETrigger.GUIDETriggerState _state = GUIDETrigger.GUIDETriggerState.None;
	internal Vector3 _value = Vector3.Zero;

	private List<GUIDETrigger> _triggerList = new();
	private int _implicitCount = 0;
	private int _explicitCount = 0;

	internal void _Initialize(GUIDEAction.GUIDEActionValueType valueType) {
		_triggerList.Clear();
		_implicitCount = 0;
		_explicitCount = 0;
		_triggerHoldThreshold = -1.0f;

		if (Triggers.Count == 0) {
			var defaultTrigger = new GUIDETriggerDown();
			defaultTrigger.ActuationThreshold = 0;
			_explicitCount = 1;
			_triggerList.Add(defaultTrigger);
			return;
		}

		Vector3 input_value = Input != null ? Input._Value : Vector3.Zero;

		foreach (var modifier in Modifiers) {
			input_value = modifier._ModifyInput(input_value, 0, valueType);
		}

		foreach (var trigger in Triggers) {
			switch (trigger._GetTriggerType()) {
				case GUIDETrigger.GUIDETriggerType.Explicit:
					_explicitCount++;
					break;
				case GUIDETrigger.GUIDETriggerType.Implicit:
					_implicitCount++;
					break;
			}
			_triggerList.Add(trigger);

			if (trigger is GUIDETriggerHold holdTrigger) {
				if (_triggerHoldThreshold == -1) {
					_triggerHoldThreshold = holdTrigger.HoldTreshold;
				}
				else {
					_triggerHoldThreshold = Math.Min(_triggerHoldThreshold, holdTrigger.HoldTreshold);
				}
			}

			trigger._LastValue = input_value;
		}
	}

	internal void _UpdateState(float delta, GUIDEAction.GUIDEActionValueType valueType) {
		Vector3 input_value = Input != null ? Input._Value : Vector3.Zero;

		foreach (var modifier in Modifiers) {
			input_value = modifier._ModifyInput(input_value, delta, valueType);
		}

		_value = input_value;

		int triggeredImplicits = 0;
		int triggeredExplicits = 0;
		int triggeredBlocked = 0;

		GUIDETrigger.GUIDETriggerState result = GUIDETrigger.GUIDETriggerState.None;

		foreach (var trigger in _triggerList) {
			var triggerResult = trigger._UpdateState(_value, delta, valueType);
			trigger._LastValue = _value;

			var triggerType = trigger._GetTriggerType();
			if (triggerResult == GUIDETrigger.GUIDETriggerState.Triggered) {
				switch (triggerType) {
					case GUIDETrigger.GUIDETriggerType.Explicit:
						triggeredExplicits++;
						break;
					case GUIDETrigger.GUIDETriggerType.Implicit:
						triggeredImplicits++;
						break;
					case GUIDETrigger.GUIDETriggerType.Blocking:
						triggeredBlocked++;
						break;
				}
			}

			if (triggerType == GUIDETrigger.GUIDETriggerType.Explicit) {
				result = (GUIDETrigger.GUIDETriggerState)Math.Max((int)result, (int)triggerResult);
			}
		}

		if (triggeredBlocked > 0) {
			_state = GUIDETrigger.GUIDETriggerState.None;
			return;
		}

		if (triggeredImplicits < _implicitCount) {
			_state = GUIDETrigger.GUIDETriggerState.None;
			return;
		}

		if (_explicitCount == 0 && _implicitCount > 0) {
			_state = GUIDETrigger.GUIDETriggerState.Triggered;
			return;
		}

		_state = result;
	}
}
